cmd/egress-cni-plugin/main.go

```diff
@@ -0,0 +1,183 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"runtime"
+	"strconv"
+
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/iptableswrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/vethwrapper"
+
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/hostipamwrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/netlinkwrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/nswrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/procsyswrapper"
+
+	"github.com/coreos/go-iptables/iptables"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	cniversion "github.com/containernetworking/cni/pkg/version"
+	"github.com/containernetworking/plugins/pkg/utils"
+)
+
+var version string
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, nil, cmdDel, cniversion.All, fmt.Sprintf("egress CNI plugin %s", version))
+}
+
+func cmdAdd(args *skel.CmdArgs) (err error) {
```

Comment:
We should move this to a `New` function in `EgressContext.go` and call that from here. We're essentially creating a new instance of egressContext and are replicating it across `cmdAdd` and `cmdDel` right now.

Reply:
I created two New to create egress context for Add and Del.